src/CommunityToolkit.Maui.Maps/Handler/Map/MapHandler.Windows.cs

```diff
@@ -366,37 +385,80 @@ void WebViewNavigationCompleted(WebView2 sender, CoreWebView2NavigationCompleted
 
 	void WebViewWebMessageReceived(WebView2 sender, CoreWebView2WebMessageReceivedEventArgs args)
 	{
-		// TODO make this better, separate events?
-		var clickedPinWebView = JsonSerializer.Deserialize<Pin>(args.WebMessageAsJson, jsonSerializerOptions);
-		var clickedPinWebViewId = clickedPinWebView?.MarkerId?.ToString();
+		var eventMessage = JsonSerializer.Deserialize<EventMessage>(args.WebMessageAsJson, jsonSerializerOptions);
 
-		if (!string.IsNullOrEmpty(clickedPinWebViewId))
+		if (eventMessage is null)
 		{
-			var clickedPin = VirtualView.Pins.SingleOrDefault(p => (p as Pin)?.Id.ToString().Equals(clickedPinWebViewId) ?? false);
-
-			clickedPin?.SendMarkerClick();
 			return;
 		}
 
-		var clickedInfoWindowWebView = JsonSerializer.Deserialize<InfoWindow>(args.WebMessageAsJson, jsonSerializerOptions);
-		var clickedInfoWindowWebViewId = clickedInfoWindowWebView?.InfoWindowMarkerId;
-
-		if (!string.IsNullOrEmpty(clickedInfoWindowWebViewId))
+		if (!Enum.TryParse<EventIdentifier>(eventMessage.Id, true, out var eventId))
 		{
-			var clickedPin = VirtualView.Pins.SingleOrDefault(p => (p as Pin)?.Id.ToString().Equals(clickedInfoWindowWebViewId) ?? false);
-
-			clickedPin?.SendInfoWindowClick();
 			return;
 		}
 
-		var mapRect = JsonSerializer.Deserialize<Bounds>(args.WebMessageAsJson, jsonSerializerOptions);
-		if (mapRect?.Center is not null)
+		switch (eventId)
 		{
-			VirtualView.VisibleRegion = new MapSpan(new Location(mapRect.Center.Latitude, mapRect.Center.Longitude),
-				mapRect.Height, mapRect.Width);
+			case EventIdentifier.BoundsChanged:
+				var mapRect = JsonSerializer.Deserialize<Bounds>(eventMessage.Payload?.ToString() ?? "", jsonSerializerOptions);
+				if (mapRect?.Center is not null)
+				{
+					VirtualView.VisibleRegion = new MapSpan(new Location(mapRect.Center.Latitude, mapRect.Center.Longitude),
+						mapRect.Height, mapRect.Width);
+				}
+				break;
+			case EventIdentifier.MapClicked:
+				var clickedLocation = JsonSerializer.Deserialize<Location>(eventMessage.Payload?.ToString() ?? "",
+					jsonSerializerOptions);
+				if (clickedLocation is not null)
+				{
+					VirtualView.Clicked(clickedLocation);
+				}
+				break;
+
+			case EventIdentifier.InfoWindowClicked:
+				var clickedInfoWindowWebView = JsonSerializer.Deserialize<InfoWindow>(eventMessage.Payload?.ToString() ?? "",
```

Comment:
Does Deserialize return null if string is empty?

Reply:
Hm thats a good test! I would expect it to indeed return null, that’s kind of what happened in the previous code as well, right?

Reply:
Good call! If you really give it an empty string it will give an exception although it won't surface, but I'll add some code to check for that first.